[Feat] Add location co-ordinates to pictures.

dev-ritik opened the pull request:
**Description (required)**

Fixes [4751](https://redmine.named-data.net/issues/4751)

Add a location checkbox to send the location data with the picture.
To implement this feature, 
 - `ACCESS_FINE_LOCATION` permission is added in the manifest file
 - Location in latitude and longitude is added in the `Exif` metadata of the image.
 - Location boolean is added to `Syncdata`
 - A new RealmObject `FilesInfoRealm` is added to store information like location for files.
 - The location is just displayed over the images as of now.

Also, bugs, probably generated over the last few commits, relating to sending files are fixed.

dev-ritik committed:
Fix friendlist error

dev-ritik committed:
Fix friendlist update on non UI thread

dev-ritik committed:
Fix extra / in data name

dev-ritik committed:
Fix viewmodel getFriend bug

dev-ritik committed:
Fix / namespace error in MemoryCache

dev-ritik committed:
Add location transfer with picture data

To implement this feature, Fine location permission is added in the manifest file, location in latitude and longitude with 4
decimal precision is added as the last 8 + 9 bytes in the unencrypted byte array, location boolean is added to syncdata.
Also, the latest SegmentFetcher from jNDN is placed with little modification and segment verification is bypassed.

dev-ritik committed:
Add FilesInfoRealm to store information about files

jrandallclark commented on app/src/main/java/memphis/myapplication/data/tasks/RttEstimator.java:
```diff
@@ -0,0 +1,171 @@
+/**
```
This file and the SegmentFetcher file shouldn't be included in npChat. Your update has been merged into the jNDN library, so we should be using that.

dev-ritik replied:
Replaced

jrandallclark commented on app/src/main/java/memphis/myapplication/utilities/FileManager.java:
```diff
@@ -208,7 +210,20 @@ public boolean saveContentToFile(Blob content, Name path) {
+        if (location) {
+            byteContent = content.getImmutableArray();
+            latitude = new String(Arrays.copyOfRange(byteContent, byteContent.length - 17, byteContent.length - 9));
+            longitude = new String(Arrays.copyOfRange(byteContent, byteContent.length - 9, byteContent.length));
```
Seeing this in action. I don't like it. I think using exif metadata is the best bet here. @agawande ?

dev-ritik replied:
Replaced this with `androidx.exifinterface.media.ExifInterface`. The coordinates are now stored in the location metadata of the image itself.

dev-ritik committed:
Use Exif for storing location coordinates

Uses Exif to store coordinates with the file instead of sending and storing it separately

dev-ritik committed:
Replaced SegmentFetcher from the library

jrandallclark commented on app/src/main/java/memphis/myapplication/UI/SelectRecipientsFragment.java:
```diff
@@ -201,31 +270,140 @@ public void onItemClick(View view, int position) {
-            @Override
-            public void onClick(View view) {
-                returnList();
+    private void work() {
```
It would be better if this function's name described what it's purpose better.

jrandallclark commented on app/src/main/java/memphis/myapplication/UI/SelectRecipientsFragment.java:
```diff
@@ -201,31 +270,140 @@ public void onItemClick(View view, int position) {
+                // If GPS enabled, get latitude/longitude using GPS Services
+                if (isGPSEnabled) {
+                    if (location == null) {
+                        Timber.i("here");
```
Should be removed?

jrandallclark commented:
I've discovered a few issues. Some of these might have been introduced in your last PR and we missed them:

1. You can highlight "Friends" in the recipients list and hit send. This causes the app to crash.

2. There's an issue somewhere with the trading of certificates. When making friends, one device gets their friend's certificate and public key, but the other device somehow gets the wrong certificate. I'm not sure exactly where the problem lies. 
Here's phone B trying unable to decrypt the symmetric key that was signed by B's public key.
`07-18 14:40:17.580 7260-7354/memphis.myapplication W/System.err: net.named_data.jndn.security.tpm.TpmBackEnd$Error: Error in TpmPrivateKey.decrypt: net.named_data.jndn.security.tpm.TpmPrivateKey$Error: Error decrypting with private key: data hash wrong
07-18 14:40:17.580 7260-7354/memphis.myapplication W/System.err:     at net.named_data.jndn.security.tpm.TpmKeyHandleMemory.doDecrypt(TpmKeyHandleMemory.java:65)
07-18 14:40:17.580 7260-7354/memphis.myapplication W/System.err:     at net.named_data.jndn.security.tpm.TpmKeyHandle.decrypt(TpmKeyHandle.java:56)
07-18 14:40:17.580 7260-7354/memphis.myapplication W/System.err:     at memphis.myapplication.psync.ConsumerManager$2.onData(ConsumerManager.java:96)`

And here's phone A unable to verify segments signed by B's public key.
`07-18 14:54:57.179 20401-21229/memphis.myapplication D/Log: 177 : FetchingTask : getUserInfo: username&PubKey: user: mb
07-18 14:54:57.180 20401-21229/memphis.myapplication D/Log: 109 : FetchingTask : fetch: KeyType: RSA
07-18 14:54:57.180 20401-21229/memphis.myapplication D/Log: 110 : FetchingTask : fetch: interest: /com/test/npChat/mb/file/storage/emulated/0/Android/data/memphis.myapplication/files/photos/2019.07.18.14.54.47.jpg
07-18 14:54:57.283 20401-21229/memphis.myapplication D/Log: 119 : FetchingTask : verifySegment: verifying segment
07-18 14:54:57.285 20401-21229/memphis.myapplication D/Log: 205 : FetchingTask : verifySignature: We've made it to verify(signature)
07-18 14:54:57.299 20401-20401/memphis.myapplication D/Log: 219 : FetchingTask : onPostExecute: Calling onPostExecute
07-18 14:54:57.299 20401-20401/memphis.myapplication D/Log: 221 : FetchingTask : onPostExecute: onPostExecute: not received
07-18 14:54:57.299 20401-20401/memphis.myapplication D/Log: 269 : FetchingTask : onPostExecute:  onError: Segment verification failed`

In this case, it seems that phone B has phone A's correct certificate. But phone A doesn't have B's.

3. GPS doesn't work for me at all. I don't receive any notification that location isn't turned on or asking for permissions. Manually adding location permissions in the settings didn't change anything, either. Also, log statements don't explain anything and need to be improved or removed.
`07-18 14:50:18.437 19508-19508/memphis.myapplication I/Log: 315 : SelectRecipientsFragmen : work: true : false
07-18 14:50:18.437 19508-19508/memphis.myapplication I/Log: 340 : SelectRecipientsFragmen : work: here
07-18 14:50:18.445 19508-19508/memphis.myapplication D/Log: 347 : SelectRecipientsFragmen : work: GPS Enabled: null
07-18 14:50:22.039 19508-19508/memphis.myapplication I/Log: 315 : SelectRecipientsFragmen : work: true : false
07-18 14:50:22.040 19508-19508/memphis.myapplication I/Log: 340 : SelectRecipientsFragmen : work: here
07-18 14:50:22.047 19508-19508/memphis.myapplication D/Log: 347 : SelectRecipientsFragmen : work: GPS Enabled: null
`

dev-ritik committed:
Fix errors in getting location with some other minor fixes

dev-ritik commented:
Fixed 1 & 3. Hopefully, fetching location must work now.

jrandallclark commented:
I get the alert asking for location permission, but then the GUI changes to "Updating..." and nothing happens. When I try to send the photo, the logs show that location that is added is 0.0 : 0.0.

Did you get different results on your device?

jrandallclark commented:
Ok, I get the coordinates on my personal phone. So it must be a device issue. My test devices are developer edition Motox's, so maybe that's the problem.

jrandallclark commented:
Instead of displaying the coordinates to the user (which would be meaningless for most people), can we use the Geocoder class (https://developer.android.com/reference/android/location/Geocoder.html) to display the city and country name instead?

jrandallclark commented:
> The latest SegmentFetcher from jNDN is placed with little modification and segment verification is bypassed.

I didn't see this before. Why was segment verification bypassed?

dev-ritik commented:
Ok, I will try to add `Geocoder` class.
I have removed the 'bypassing' before pushing the fix (I did it because I was getting the same error as you mentioned above, but I left it, for the time being, considering it not to be related with the main objective of the PR).

jrandallclark commented:
Ok. The validation issue is odd. User's get their friend's certificates by scanning their QR code. So I'm not sure where the problem is there.

dev-ritik committed:
Fix segment validation error

dev-ritik committed:
Use Geocoder to get address from Location

dev-ritik commented:
@jrandallclark PTAL.

jrandallclark commented on app/src/main/java/memphis/myapplication/data/Common.java:
```diff
@@ -143,14 +143,14 @@ public void run() {
             meta_info.setType(ContentType.BLOB);
             // not sure what is a good freshness period
             meta_info.setFreshnessPeriod(90000);
+            datas.add(data);
```
Why move this? Your code never adds the final block ID to the last segment, only the first. This will be replaced by the SegmentPublisher in the future anyways, but should still be fixed.

dev-ritik replied:
I was getting an error in cases when the data to packetize was less than 8000 bytes long. In those cases, only 1 packet would be formed. Accessing the first object in the ArrayList before adding it resulted in an error.
I would fix the new issue by adding `MetaInfo` to the last element of the ArrayList as well.

jrandallclark replied:
Not sure why that only started happening only now. Either way, your solution works for the moment and this function is only temporary.

jrandallclark reviewed (changes requested):
Other than my question about the packetizing function, everything seems to be working fine.

dev-ritik committed:
Add MetaInfo to the last segment

jrandallclark commented:
Everything looks good. Great job!
